examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,577 @@
+            ),
+            layers.Dense(representation_dim),
+        ]
+    )
+    return encoder
+
+
+"""
+### Implement the unsupervised contrastive loss
+"""
+
+
+class ContrastiveLoss(keras.losses.Loss):
+    def __init__(self, temperature=1.0, l2_normalize=False):
+        super(ContrastiveLoss, self).__init__()
+        self.temperature = temperature
+        self.l2_normalize = l2_normalize
+
+    def __call__(self, labels, feature_vectors, sample_weight=None):
+        # The shape of feature_vectors is [num_augmentations * batch_size, projection_units].
+        # Convert labels to an identiy matrix of shape [batch_size, batch_size].
+        labels = tf.eye(tf.shape(labels)[0])
+        if self.l2_normalize:
+            feature_vectors = tf.math.l2_normalize(feature_vectors, -1)
+        # The logits shape is [num_augmentations * batch_size, num_augmentations * batch_size].
+        logits = (
+            tf.linalg.matmul(feature_vectors, feature_vectors, transpose_b=True)
+            / self.temperature
+        )
+        # Apply log-max trick for numerical stability.
+        logits_max = tf.math.reduce_max(logits, axis=1)
+        logits = logits - logits_max
+        # Compute num_augmentations.
+        num_augmentations = tf.shape(feature_vectors)[0] // tf.shape(labels)[0]
+        # The shape of targets is [num_augmentations * batch_size, num_augmentations * batch_size].
+        targets = tf.tile(labels, [num_augmentations, num_augmentations])
+        # Compute cross entropy loss
+        loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        return loss
+
+
+"""
+### Implement representation learner
+"""
+
+
+def create_representation_learner(encoder, num_augumentations, projection_units):
+
+    inputs = keras.Input(shape=input_shape)
+    # Preprocess the input images.
+    preprocessed = data_preprocessing(inputs)
+    # Create augmented versions of the images.
+    augmented = []
+    for _ in range(num_augumentations):
+        augmented.append(data_augmentation(preprocessed))
+    augmented = layers.Concatenate(axis=0)(augmented)
+    # Generate embedding representations of the images.
+    features = encoder(augmented)
+    # Add a projection head.
+    features = layers.Dropout(rate=0.1)(features)
+    x = layers.Dense(units=projection_units, use_bias=False)(features)
+    x = layers.BatchNormalization()(x)
+    outputs = layers.ReLU()(x)
+    # Create the model.
+    model = keras.Model(inputs=inputs, outputs=outputs, name="representation_learner")
+    return model
+
+
+"""
+### Train the model
+"""
+
+encoder = create_encoder(representation_dim)
+representation_learner = create_representation_learner(
+    encoder, num_augumentations, projection_units
+)
+representation_learner.summary()
+
+# Since this is self-supervised learning, we don't
+# use the y_data as our labels.
+labels = tf.ones(shape=(x_data.shape[0]))
```

Comment:
This is a hack. For this reason I would suggest instead taking the approach of subclass `Model` and writing a custom `train_step`.

Reply:
I agree. However, I really find it much simpler than writing a custom `train_step`

Maybe the keras APIs (around fit and loss) can be updated to allow `y` parameter to be `None`? This is useful in self-supervised and unsupervised use case, where there is no targets, and the loss is computed only based on the input data `x`?

Reply:
The default `compile`/`fit` flow is for supervised learning. If you're not doing supervised learning, then you have to customize it. And that means either writing a training loop from scratch, or writing a custom `train_step` (which is simpler).

You find the current version much simpler because it represents less code, but here's the thing: if you want to customize things further, you get much less flexibility. Whereas if you have a custom `train_step`, then modifying things further is pretty easy. You don't fall off a cliff. That's the problem with hacks: they work only until you want to start changing things. They're not flexible.

Further, precisely for this reason (increased flexibility), we need to teach users about the `train_step` overriding workflow.

Happy to guide you with implementing that.

Reply:
@fchollet Thank you for your feedback. It would be great if I can follow your guidance in implementing this.